src/source/rtl_text_plugin_main_thread.ts

```
@@ -1,60 +1,85 @@
-        if (!deferred) {
-            //Start downloading the plugin immediately if not intending to lazy-load
-            await this._downloadRTLTextPlugin();
+
+        this.url = browser.resolveURL(url);
+        if (!this.url) {
+            throw new Error(`requested url ${url} is invalid`);
         }
-    }
+        if (this.status === 'unavailable') {
+
+            // from initial state:
+            if (deferred) {
+
+                this.status = 'deferred';
+                // fire and forget: in this case it does not need wait for the broadcasting result
+                // it is important to sync the deferred status once because
+                // symbol_bucket will be checking it in worker
+                this._syncState(this.status);
 
-    async _downloadRTLTextPlugin() {
-        if (this.pluginStatus !== 'deferred' || !this.pluginURL) {
-            throw new Error('rtl-text-plugin cannot be downloaded unless a pluginURL is specified');
+            } else {
+                return this._requestImport();
+            }
+
+        } else if (this.status === 'requested') {
+            return this._requestImport();
         }
-        try {
-            this.pluginStatus = 'loading';
-            await this._sendPluginStateToWorker();
-            await getArrayBuffer({url: this.pluginURL}, new AbortController());
-            this.pluginStatus = 'loaded';
-        } catch {
-            this.pluginStatus = 'error';
+    }
+
+    /** Send a message to worker which will import the RTL plugin script */
+    async _requestImport() : Promise<void> {
+        const workerResults = await this._syncState('loading');
+
+        // expect all of them to be 'loaded'
+        const expectedStatus = 'loaded';
+        const failedToLoadWorkerExist: boolean = workerResults.some((workerResult) => {
+            return workerResult.pluginStatus !== expectedStatus;
+        });
+
+        if (failedToLoadWorkerExist) {
```

I believe that in a case of error, since you are throwing in the worker, this code won't be reached but instead, you'll need to try-catch the `_syncState` call.
Feel free to verify me.

---

For clear error cases (networking, bad script, failed to parse ..), worker will throw an exception. On console it looks like:
<code>Uncaught (in promise) Error: RTL Text Plugin failed to import scripts from https://somehost/foo.js</code>
or
<code>uncaught (in promise) DOMException: Failed to execute 'importScripts' on 'WorkerGlobalScope': The script at 'https://foo/rtl/foo.js' failed to load.</code>

Catching or not catching in <code>_syncState</code> would not make any difference, because it can only set error state and throw again, not swallowing the error. In reality it does not help dev or end users in anyway.

<code>return workerResult.pluginStatus !== expectedStatus;</code> is NOT intended to handle anything above.  It is only looking for anything that is not 'loaded', so just a "catch all" to process anything unforeseeable in messaging and worker, which means for whatever unknown reason, some or all workers failed to reach 'loaded' state.

If we take out this part, it means main will fire RTLPluginLoaded without checking the result from worker, the next dev to debug this code will be left in dark and very confused of that what happened.

---

In any case of a worker error, the worker will throw, but the main thread state won't be error, which I don't think is a good idea.
If there is any problem in the wroker, it will throw, which is great, so this code is dead code, unless you can tell me how to reach it...

I would consider addidng tests that involve both the main thread and the worker code to make sure we can reproduce the scenarios we are talking about, as the current tests mock half the code at every scenario, and the flows we are talking about are more in the integration realm...

---

Moved catch to _syncState now. I have not add/modified any test cases --- let's agree on the code logic first.

---

This looks good now. thanks!
You can use `try { ... } catch` instead of throwing inside the `catch` callback as I'm not sure it will throw to the "right place".

---

.catch block is correct. Experimented with <code>try/catch</code> and it did not catch anything .

---

That doesn't make any sense, try catch with await and .catch are equivalent.

---

> That doesn't make any sense, try catch with await and .catch are equivalent.

syncState is returning a promise, not doing await because:
- "setRTLTextPlugin" is using syncState for 'deferred' status and that is "fire and forget" and no need for await. A minor perf improvement
- "requestImport" is doing await on the promise returned by syncState
Both cases, .catch will capture and everything behaves correctly

---

This isn't critical. Let's finish with this PR...